andi/andi.py

```diff
@@ -1,62 +1,318 @@
 # -*- coding: utf-8 -*-
+from collections import OrderedDict
 from typing import (
     Dict, List, Optional, Type, Callable, Union, Container,
-    get_type_hints,
-)
+    get_type_hints, Tuple, cast, MutableMapping)
 
-from andi.typeutils import get_union_args, is_union, get_globalns
+from andi.typeutils import (
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanMapping = MutableMapping[Type, Dict[str, Type]]
+
+
+class Plan(PlanMapping):
```

Comment:
I wonder how to make this more clean. There are 2 problems with the current implementation:

1) inheriting from a type; I'm not sure it is an intended usage of typing module
2) copy-paste for magic methods.

It seems (2) can be fixed by either inheriting from OrderedDict, or by removing methods which are not needed, and maybe making an attribute with OrderedDict public.

Reply:
> inheriting from a type; I'm not sure it is an intended usage of typing module

I think it makes sense to inherit from typing. They are the same than the `collections` ABC base classes but with strict typing. See https://github.com/python/typing/issues/203#issuecomment-250262040 

> It seems (2) can be fixed by either inheriting from OrderedDict, 

The implementation will be much simpler if we extend OrderedDict, but we would lose the types as it is not possible to do:
```py
class Plan(OrderedDict[Type, Dict[str, Type]])
```

Instead, we would have to do:
```py
class Plan(OrderedDict)
```
The implementation would be much simpler (no magic types would be required), but functionally it would be worse as it wouldn't be strongly typed.

> or by removing methods which are not needed, and maybe making an attribute with OrderedDict public.

I have added all the methods required from as stated in https://docs.python.org/3/library/collections.abc.html. See the following excerpt:
![image](https://user-images.githubusercontent.com/895720/76006508-7d01cf80-5f04-11ea-83af-8289eb9f7aca.png)

So that Plan is just delegating all its `MutableMapping` functionality to the underlying OrderedDict.

Well, so I'm not convinced of your alternative proposals in this case.

Reply:
Mmm, maybe the following could be possible:
```py
class Plan(OrderedDict, MutableMapping[Type, Dict[str, Type]]):
```
Tomorrow I'll have a look. This would make all the magic methods obsolete.

Reply:
I was sure you considered these options :)

However, why is MutableMapping interface needed here? What do you think about something like that?

```py
class Plan:
    def __init__(self, steps: OrderedDict):
        self.steps = steps  # or should we call it .plan?

    @property
    def dependencies(self) -> List[Tuple[Callable, Dict[str, Type]]]:       
        # a list is returned, because this allows to avoid .items() 
        # call in build function, and also because it makes it clear
        # order is important
    
    @property
    def final_arguments(self) -> Dict[str, Type]:
        # ....
```

`_plan` function can then use OrderedDict internally, to build _plan_seq.

Reply:
I'm working now with something like that:

```py
PlanList = List[Tuple[Callable, Dict[str, Callable]]]

class Plan(PlanList):

    @property
    def dependencies(self) -> PlanList:
        return self[:-1]

    @property
    def final_arguments(self) -> Dict[str, Type]:
        _, params = self[-1]
        return params
```

I'm going to see how the code looks like with it.

Reply:
So finally I like your idea of just using lists of steps instead of a full ordered dictionary. I have implemented it here: https://github.com/scrapinghub/andi/pull/8/commits/2945a6997c920deef3ca0cd6782945c397789cdf

Besides, build an ordered dict from them is very easy, just `OrderedDict(plan)`.